### PR TITLE
p2p: trusted peers, dial throttling

### PR DESCRIPTION
**Pull request by karalabe**
This PR adds support for privileged trusted peers (issue #831):

 - A list of enode URLs (~~either as cmd literal or~~ *internal* json file) may be set as trusted nodes
 - Trusted nodes are privileged: they get connected/accepted even above the maximum peer count
 - Connections to trusted nodes are reattempted every once in a while (15 sec for now)

**Note**: `suggestPeer` has been dropped in favor of `trustPeer`. Previously suggest was just a one shot dial suggestion (i.e. "pls pls pls connect") but if the dial failed, it was never retried. The new semantics was extended to retry connecting as long as the node is live.

A few nice to have features:

 - [x] Update the dialer to maintain a list of recent failures to prevent constant redialing (i.e. have a cooldown of 30~~-60~~ seconds before reattempting a previously failed peer).

Needed before merge:

 - [x] Test to verify that trusted peers indeed skip the normal evacuation checks.
 - [x] Tests to verify inbound and outbound connection throttling after failures.

### CODE REVIEW

File: p2p/server.go [hunk 1]
```diff
@@ -472,16 +567,29 @@ func (srv *Server) addPeer(id discover.NodeID, p *Peer) (bool, DiscReason) {
...
 	case srv.peers[id] != nil:
 		return false, DiscAlreadyConnected
+
 	case id == srv.ntab.Self().ID:
 		return false, DiscSelf
+
+	case srv.throttle[id]:
+		return false, DiscThrottled
+
 	default:
```
Comment (fjl) [inline p2p/server.go]: Please don't add blank lines in this function. They are not necessary. I'm fine with doing this for select/switch with large case bodies, but not here.

File: p2p/server.go [hunk 2]
```diff
@@ -95,20 +101,24 @@ type Server struct {
 
 	ourHandshake *protoHandshake
 
-	lock    sync.RWMutex // protects running and peers
-	running bool
-	peers   map[discover.NodeID]*Peer
+	lock sync.RWMutex // protects running and peers
```
Comment (fjl) [inline p2p/server.go]: Style: please put the lock directly above the fields it protects without adding a blank line.

File: p2p/handshake.go [hunk 3]
```diff
@@ -70,21 +70,21 @@ type protoHandshake struct {
...
 
-func setupInboundConn(fd net.Conn, prv *ecdsa.PrivateKey, our *protoHandshake, atcap bool) (*conn, error) {
+func setupInboundConn(fd net.Conn, prv *ecdsa.PrivateKey, our *protoHandshake, atcap bool, trust map[discover.NodeID]bool) (*conn, error) {
 	secrets, err := receiverEncHandshake(fd, prv, nil)
 	if err != nil {
 		return nil, fmt.Errorf("encryption handshake failed: %v", err)
 	}
 	rw := newRlpxFrameRW(fd, secrets)
-	if atcap {
+	if atcap && !trust[secrets.RemoteID] {
```
Comment (fjl) [inline p2p/handshake.go]: This is not a good solution. Having to tunnel `atcap` into these functions is bad enough.
It would be better if you could ensure that `atcap` is always false for trusted dials.

File: p2p/peer_error.go [hunk 4]
```diff
@@ -73,6 +73,7 @@ const (
 	DiscSelf
 	DiscReadTimeout
 	DiscSubprotocolError
+	DiscThrottled
```
Comment (fjl) [inline p2p/peer_error.go]: You cannot add reasons to this list because these values are part of the protocol definition.

File: p2p/server.go [hunk 5]
```diff
@@ -323,6 +345,45 @@ func (srv *Server) listenLoop() {
 	}
 }
 
+// trustLoop is responsible for periodically checking that trusted connections
+// are actually live, and requests dialing if not.
+func (srv *Server) trustLoop() {
```
Comment (fjl) [inline p2p/server.go]: Please call this `trustedNodesLoop`.

File: p2p/server.go [hunk 6]
```diff
@@ -323,6 +345,45 @@ func (srv *Server) listenLoop() {
...
+			for _, node := range needed {
+				glog.V(logger.Debug).Infof("Dialing trusted peer %v", node)
+				select {
+				case srv.trustDial <- node:
+					// Ok, dialing
+
+				case <-srv.quit:
+					// Terminating, return
+					return
+				}
```
Comment (fjl) [inline p2p/server.go]: Style: comments, blank lines are not necessary in this select statement. Reders are supposed to know how select works.

File: p2p/server.go [hunk 7]
```diff
@@ -131,10 +141,12 @@ func (srv *Server) PeerCount() int {
 	return n
 }
 
-// SuggestPeer creates a connection to the given Node if it
-// is not already connected.
-func (srv *Server) SuggestPeer(n *discover.Node) {
-	srv.peerConnect <- n
+// TrustPeer inserts a node into the list of privileged nodes.
+func (srv *Server) TrustPeer(node *discover.Node) {
```
Comment (fjl) [inline p2p/server.go]: This function should be called `AddPeer`. The comment should say:

```go
// AddPeer connects to the given node and maintains the connection until
// the server is shut down.
```

File: p2p/server.go [hunk 8]
```diff
@@ -323,6 +345,45 @@ func (srv *Server) listenLoop() {
...
+// trustLoop is responsible for periodically checking that trusted connections
+// are actually live, and requests dialing if not.
+func (srv *Server) trustLoop() {
+	// Create a ticker for verifying trusted connections
+	tick := time.Tick(trustedPeerCheckInterval)
+
+	for {
+		select {
+		case <-srv.quit:
+			// Termination requested, simple return
```
Comment (fjl) [inline p2p/server.go]: Style: no comment

File: eth/backend.go [hunk 9]
```diff
@@ -96,6 +102,43 @@ func (cfg *Config) parseBootNodes() []*discover.Node {
 	return ns
 }
 
+// parseTrustedNodes parses a list of discovery node URLs either given literally,
+// or loaded from a .json file.
+func (cfg *Config) parseTrustedNodes() []*discover.Node {
```
Comment (fjl) [inline eth/backend.go]: I'm thinking about adding an `UnmarshalJSON` to `discover.Node`.

File: eth/backend.go [hunk 10]
```diff
@@ -96,6 +102,43 @@ func (cfg *Config) parseBootNodes() []*discover.Node {
...
+	if err != nil {
+		glog.V(logger.Error).Infof("Failed to access trusted nodes: %v", err)
+		return nil
+	}
+	nodelist := []string{}
+	if err := json.Unmarshal(blob, &nodelist); err != nil {
+		glog.V(logger.Error).Infof("Failed to load trusted nodes: %v", err)
+		return nil
+	}
+	fmt.Println("nodes", nodelist)
```
Comment (fjl) [inline eth/backend.go]: This needs to go away.

File: cmd/geth/main.go [hunk 11]
```diff
@@ -232,7 +232,7 @@ JavaScript API. See https://github.com/ethereum/go-ethereum/wiki/Javascipt-Conso
 		utils.IdentityFlag,
 		utils.UnlockedAccountFlag,
 		utils.PasswordFileFlag,
-		utils.BootnodesFlag,
+		utils.BootNodesFlag,
```
Comment (fjl) [inline cmd/geth/main.go]: This change is unnecessary. Please don't make unrelated changes like this.

File: cmd/geth/admin.go [hunk 12]
```diff
@@ -25,7 +25,7 @@ func (js *jsre) adminBindings() {
 	js.re.Set("admin", struct{}{})
 	t, _ := js.re.Get("admin")
 	admin := t.Object()
-	admin.Set("suggestPeer", js.suggestPeer)
+	admin.Set("trustPeer", js.trustPeer)
```
Comment (fjl) [inline cmd/geth/admin.go]: `addPeer`.

File: p2p/handshake.go (hunk 3 above)
Comment (karalabe) [inline p2p/handshake.go]: The problem is that I need secrets.RemoteID to check if the peer is trusted, so I need the encryption handshake to go through before I can filter. Btw, it's not the real map, only a copy.

File: p2p/peer_error.go (hunk 4 above)
Comment (karalabe) [inline p2p/peer_error.go]: Ah, wanted to as about that. What error should I return in the case of throttling then?

**Comment by obscuren** [comment]
Please split up this PR in 2 separate PRs

**Comment by karalabe** [comment]
The tursted peers were cherry picked into this PR: https://github.com/ethereum/go-ethereum/pull/836

**Comment by karalabe** [comment]
Connection throttling separated into https://github.com/ethereum/go-ethereum/pull/839

Closing this PR.
